### ISSUE TITLE
build-package fontconfig-2.13.1-4 broken

### ISSUE THREAD

**Issue by kallisti5** [haikuports/haikuports#4266]
The fontconfig package is broken... looking for lib:libuuid 1.3.0 instead of 1.0.3??

```
problem 71: nothing provides lib:libuuid>=1.3.0 needed by fontconfig-2.13.1-4
+       libuuid-1.0.3-6
```

This is causing some major issues for the builds.  Could someone fix it ASAP?

**Comment by kallisti5** [comment]
@korli     Seems to be due to this?


https://github.com/haikuports/haikuports/blob/master/media-libs/fontconfig/fontconfig-2.13.1.recipe#L65

1.3 == 1.3.0  vs 1.0.3?

https://github.com/haikuports/haikuports/commit/6d33b5ffe215e4a5e3027be6af89525ff0f30e5b


looks like "1.3" was introduced here:
https://github.com/haikuports/haikuports/commit/500223359fa407cd2cc8f22c44292e34fd73938e

How is this working?

**Comment by kallisti5** [comment]
wait.. these depend on https://github.com/haikuports/haikuports/blob/master/sys-apps/util-linux/util_linux-2.34.recipe#L50 now instead of libuuid?

**Comment by kallisti5** [comment]
So.. The solution seems to be disabling libuuid in util_linux and moving back to the (better licensed) libuuid.

**Comment by korli** [comment]
Please provide more context to your actual problem.

**Comment by kallisti5** [comment]
Ok... some basics.

 1. We have a set of build-packages which are used during the build. These are the packages pre-installed into Haiku, and are generally required early in the process.
 2. We update these special build packages once and a while via a complex (and dumb) hardlinking process.
 3. Compiling haiku with the updated build packages is failing due to libuuid > 1.3 missing for fontconfig (a core build package)
 4. fontconfig now depends on util_linux (and all of the other misc crap it installs) due to libuuid >1.3 coming from there.
 5. We need to ship util_linux (and all of it's dependencies) with Haiku and integrate it all into our build.

Solutions:
 1. Break util_linux into util_linux_libuuid and friends
 2. Make fontconfig use libuuid instead of util_linux_libuuid
 3. Revert to using (the better BSD licensed) libuuid 1.x

**Comment by kallisti5** [comment]
Keep in mind this doesn't even begin to address the bootstrap process where we build "haiku software" under Linux.  Not sure bootstrapping util_linux to get libuuid to get fontconfig is ideal.

**Comment by kallisti5** [comment]
@waddlesplash I wonder if we could add a special flag to build packages? hardlink_packages could auto-populate based on that and it would serve as a warning to "be gentle to this one" :P

**Comment by korli** [comment]
It's not the first time dependencies of system packages evolve, so why make this a blocker bug?

**Comment by kallisti5** [comment]
It's blocking a large amount of work refactoring our EFI code :-)  Updating these build packages is also a pre-requirement for R1 Beta2

**Comment by korli** [comment]
@waddlesplash seems in capacity here.
https://git.haiku-os.org/haiku/commit/build/jam/repositories/HaikuPorts/x86_64?id=54e53dabfc4acd2530555c717069f894fc2e6998

**Comment by waddlesplash** [comment]
Why am I being assigned? If this was just an issue of adding one package, kallisti5 would have done it. The issue here is that util_linux itself apparently has dependencies we don't yet include, and a large number of the commands in it don't work (according to kallisti5 at least, I didn't check.)

So, we should break out libuuid into a separate package again, so that we don't have to add unnecessary stuff to the build (and thus default install) packages set. Hence why there is a ticket here.

**Comment by kallisti5** [comment]
@korli I don't mind doing the leg work... but given the number of possible solutions, I don't understand why we're now falling back to libuuid from util_linux... is the code "better"? is there something wrong with bsd's implementation? is there some other reason?  None of this is documented anywhere.

My gut reaction with what's documented is to "revert those commits building libuuid from util_linux" and all the odd > 1.3  logic.  I need to understand the reasoning before I go making changes to try and fix it.

**Comment by kallisti5** [comment]
> and a large number of the commands in it don't work (according to kallisti5 at least, I didn't check.)

@waddlesplash I see comments about util_linux being "stubs" in #2525.. but it looks like those concerns were addressed in the PR? Not sure.

**Comment by korli** [comment]
> Ship util_linux and all of it's dependencies blowing out our image size
what is this actually? dependencies are "lib:libintl, lib:libncursesw, lib:libz.
If you want a package with util_linux_libs, then ask.
In the mean time, if this is a blocker, update the build packages, increase the image size.

**Comment by pulkomandy** [comment]
We can update the bootstrap dependencies indeed, but it seems annoying to import the whole util_linux there when we could use just libuuid, for the two already mentionned reasons:
1) it's bigger, and probably more complex to cross-build
2) it is under the GPL, which we'd rather avoid for core Haiku dependencies

So, there should be an option to use the BSD libuuid, unless there is a strong reason to switch to the util_linux one. So far the only reason I have seen is "util_linux is newer", but that on its own does not seem to counterbalance the two problems mentionned.

**Comment by kallisti5** [comment]
ok.. looking here, Scientific linux builds "libuuid" from util_linux.

http://rpm.pbone.net/index.php3/stat/4/idpl/16848315/dir/scientific_linux_6/com/libuuid-2.17.2-12.el6.x86_64.rpm.html

It appears libuuid "moved" into util_linux?

My vote is we make the libuuid package build from the util_linux sources.   I see similar complaints mentions here:
https://stackoverflow.com/questions/22359190/how-to-compile-only-uuid-from-linux-utils

That will (hopefully) preserve the more friendly BSD license, make things a lot simpler, and reduce the footprint for  things  that require libuuid?

**Comment by korli** [comment]
libuuid in the util_linux package is BSD licensed. Like LLVM and some other libraries they don't need to be built in the bootstrapping process. 
Sure one can split libuuid in a util_linux_libuuid subpackage. IMO this looks like "nice to have" and not like a blocker bug.

**Comment by kallisti5** [comment]
ok. after some investigation i'd decided on the unthinkable action of forking libuuid from util_linux.

My reasoning is as follows:
  * The current libuuid in util_linux is a mix of LGPL and BSD linked together (and misidentified as BSD in Fedora / other operating systems)
  * The codebase for libuuid is pretty small
  * In a few hours of work i've already made some *huge* improvements to it while keeping compatibility and removing the LGPL code.

https://gitlab.com/kallisti5/libuuid

The tests are validating, and this one is 100% BSD licensed .

**Comment by kallisti5** [comment]
Builds work under Linux / Haiku.  Getting ready to ship first version.
```
~/libuuid> scons
scons: Reading SConscript files ...
Checking for C function usleep()... (cached) yes
Checking for C function nanosleep()... (cached) yes
Checking for C header file unistd.h... (cached) yes
Checking for C header file stdlib.h... (cached) yes
Checking for C header file sys/time.h... (cached) yes
Checking for C header file sys/file.h... (cached) yes
Checking for C header file sys/ioctl.h... (cached) yes
Checking for C header file sys/un.h... (cached) yes
Checking for C header file sys/sockio.h... (cached) yes
Checking for C header file net/if.h... (cached) yes
Checking for C header file netinet/in.h... (cached) yes
Checking for C header file net/if_dl.h... (cached) yes
Checking for C header file sys/syscall.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
gcc -o src/clear.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/clear.c
gcc -o src/compare.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/compare.c
gcc -o src/copy.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/copy.c
gcc -o src/gen_uuid.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/gen_uuid.c
gcc -o src/isnull.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/isnull.c
gcc -o src/pack.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/pack.c
gcc -o src/parse.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/parse.c
gcc -o src/predefined.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/predefined.c
gcc -o src/unpack.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/unpack.c
gcc -o src/unparse.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/unparse.c
gcc -o src/uuid_time.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/uuid_time.c
gcc -o src/md5.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/md5.c
gcc -o src/sha1.os -c -fPIC -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/sha1.c
gcc -o src/libuuid.so.1.3.0 -shared -Wl,-Bsymbolic -Wl,-soname=libuuid.so.1 src/clear.os src/compare.os src/copy.os src/gen_uuid.os src/isnull.os src/pack.os src/parse.os src/predefined.os src/unpack.os src/unparse.os src/uuid_time.os src/md5.os src/sha1.os
Create symlinks for: 'src/libuuid.so.1.3.0': 'src/libuuid.so.1'->'src/libuuid.so.1.3.0', 'src/libuuid.so'->'src/libuuid.so.1.3.0'
gcc -o src/test_uuid.o -c -DHAVE_USLEEP -DHAVE_NANOSLEEP -DHAVE_UNISTD_H -DHAVE_STDLIB_H -DHAVE_SYS_TIME_H -DHAVE_SYS_FILE_H -DHAVE_SYS_IOCTL_H -DHAVE_SYS_UN_H -DHAVE_SYS_SOCKIO_H -DHAVE_NET_IF_H -DHAVE_NETINET_IN_H -DHAVE_NET_IF_DL_H -Iincludes src/test_uuid.c
gcc -o src/test_uuid src/test_uuid.o -Lsrc src/libuuid.so.1.3.0
scons: done building targets.
~/libuuid> ./src/test_uuid 
Testing UUID syntax...
84949cc5-4701-4a84-895b-354c584a981b is valid, OK
84949CC5-4701-4A84-895B-354C584A981B is valid, OK
84949cc5-4701-4a84-895b-354c584a981bc is invalid, OK
84949cc5-4701-4a84-895b-354c584a981 is invalid, OK
84949cc5x4701-4a84-895b-354c584a981b is invalid, OK
84949cc504701-4a84-895b-354c584a981b is invalid, OK
84949cc5-470104a84-895b-354c584a981b is invalid, OK
84949cc5-4701-4a840895b-354c584a981b is invalid, OK
84949cc5-4701-4a84-895b0354c584a981b is invalid, OK
g4949cc5-4701-4a84-895b-354c584a981b is invalid, OK
84949cc5-4701-4a84-895b-354c584a981g is invalid, OK
00000000-0000-0000-0000-000000000000 is valid, OK
01234567-89ab-cdef-0134-567890abcedf is valid, OK
ffffffff-ffff-ffff-ffff-ffffffffffff is valid, OK
Testing uuid_generate_random...
d3cadd2c-7433-4523-baa9-7ff44230d983 is valid, OK
48bdec76-85e4-44de-bd16-b39fdaf523df is valid, OK
8e060e2b-8450-4f18-bd9c-6ce6bbb88006 is valid, OK
Testing uuid_generate...
d6cebaf1-1e4c-48a9-839d-9507d70a5970 is valid, OK
6036fa87-7a19-47d6-8fab-ebfe72485c99 is valid, OK
78898b56-5a64-40fe-9c1c-a2f38304de5d is valid, OK
```

**Comment by kallisti5** [comment]
resolved via libuuid-1.3.0 port.
